### ISSUE TITLE
머딥 - 21d (tensorflow)

### ISSUE THREAD

**Issue by vmtmxmf5** [vmtmxmf5/Python-Basics#58]
# 목차
1. Tensorflow 기초
2. Keras DNN
3. Keras DNN - Sequential 1
4. Keras DNN - Sequential 2

## Tensorflow 기초
행렬 계산은 GPU가 많이 좌우한다 (글픽카드, ram)
```
import numpy as np
import tensorflow as tf
```

- tf.constant()
list, tuple, ndarray 를 tensor 로 변환
tuple 처럼 **데이터 수정이 불가능** (vari.만 수정가능)
```sage
# list
tf.constant([1,2,3])
# tuple
tf.constant(((1,2,3),
             (4,5,6)))
# array
arr = np.array([1,2,3])
tf.constant(arr)

tensor = tf.constant([1,2,3])
tensor[0] = 1 # error
```

- tensor 정보 확인하기
```sage
tensor.shape
tensor.dtype
```

- tensor dtype 바꾸기
```sage
tf.cast(tensor, dtype=tf.uint8)
```

- tensor 를 array 로 바꾸기
```sage
tensor.numpy()
np.array(tensor)
```

- 분포 생성
```sage
tf.random.normal([2, 4, 3])
tf.random.uniform([3, 3])
```

## Keras DNN
> **모델 구성 -> 컴파일(최적화, 평가지표) -> 하이퍼파라미터 -> 훈련 시작!**

```
from tensorflow.keras import datasets

mnist = datasets.mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train.shape, y_train.shape
X_train = X_train / 255. 
X_test = X_test / 255.
```

레이어는 tf.keras.layers 패키지에 다 들어있음
> Input Layer, Flatten Layer, Dense Layer(=Affine), Activation Layer

### 모델 구성
```sage
from tensorflow.keras.layers import Dense, Flatten, Activation, Input

inp = Input(shape=(28,28))
net = Flatten()(inp)

net = Dense(512)(net)
net = Activation('relu')(net)

net = Dense(256)(net)
net = Activation('relu')(net)

net = Dense(10)(net)
net = Activation('softmax')(net)

model = tf.keras.Model(inputs=inp, outputs=net, name='BASIC_MNIST')
model.summary()
```
# of parmeter : 784 * 512 + 512
# of parmeter : 785 * 512

### 최적화
- Loss function
출력층 뉴런이 1개 + mse = 이진분류
출력층 뉴런이 2개 + bianary_cee = 이진분류
출력층 뉴런은 class개수 + categorical_cee = 다중분류

**categorical_crossentropy**
Label & OH : categorical_crossentropy
Label & ~OH : sparse_categorical_crossentropy
> y[np.arange(batchsize), t] 원리

- optimization function
sgd : tf.keras.optimizers.SGD()
rmsprops : tf.keras.optimizers.RMSprop()
adam : tf.keras.optimizers.Adam()

- metrics
테스트세트 평가방법
분류문제에서 loss 를 테스트세트 평가지표로는 안 씀. 걍 훈련용임.
==> 훈련되기 위한 내용을 갱신하기 위해 loss 를 구함. 이게 loss 사용목적임
==> metric 에서 mse 쓰는 건 mse라고 하지 손실함수라고 하지 않는다. 
(딥러닝 분류문제에서는 CEE를 평가지표로 잘 안쓰지만, 회귀문제에서는 rmse를 평가지표로 쓸 수 있음)

```sage
y_train.shape # ~OH
loss_func = tf.keras.losses.sparse_categorical_crossentropy

optm = tf.keras.optimizers.Adam()

metrics = ['accuracy']
metrics = ['acc']
metrics = [tf.keras.metrics.Accuracy()]
```

### 모델 컴파일

```sage
model.compile(
    optimizer=optm,
    loss=loss_func,
    metrics=metrics
    )
```

### 하이퍼 파라미터 설정
- 에폭 회수(num_epochs)
- 배치크기(batch_size)

```sage
num_epochs = 10
batch_size = 32
```

### 훈련 시작
```sage
model.fit(
    X_train,
    t_train,
    batch_size=batch_size,
    epochs=num_epochs,
    shuffle=True,
    validation_split=0.2,
    )
```

## Keras Sequential Modeling
= OrderedDict
= 리스트 안에 레이어를 엮어준다

- 주의! **input_shape**를 반드시 만들어야 한다

준비
```
import numpy as np
import tensorflow as tf
from tensorflow.keras import datasets
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.models import Sequential
from sklearn.preprocessing import OneHotEncoder

mnist = datasets.mnist
(X_train, y_train), (X_test, t_test) = mnist.load_data()

y_train_oh = tf.one_hot(y_train, 10)
y_test_oh = tf.one_hot(t_test, 10)
```

```sage
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(512, activation='relu'),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(10, activation='relu'),
    ])
model.summary()

model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['acc'])

model.fit(X_train,
          y_train,
          validation_data=(X_test, y_test),
          epochs=15,
          batch_size=32)
```
